### DISCUSSION TITLE
Forum for libpeer queries - Q about using TURN service

### DISCUSSION THREAD

**Discussion by tpbedford-oomdata** [sepfy/libpeer#71] [General]
Hi, I've successfully utilised libpeer in an ESP32 project to get audio streaming working between PC and ESP32, mostly just reconfiguring the MJPEG example.

I tried specifying TURN service (eturnal; hosted on my own server) in the ESP32 and successfully generated a user/password combo, but it didn't establish ICE to the peer. STUN works just fine as the peers can NAT through to each other without issue.

So, I wondered if anyone else has this working or if there's a discussion forum I can head to. Cheers.

**Comment by sepfy** [comment]
Hello,
Ideally, libpeer can support TURN server. May i know your PeerConfiguration?
i don't think there is a discussion forum for now. haha

**Reply by tpbedford-oomdata** [reply]
Hi, I appreciate the response! My PeerConfiguration:

```
PeerConfiguration config = {
        .ice_servers = {
            { .urls = "turn:<myserver>:3478, .username = "1735686000", .credential = "XXXX" },
            //{ .urls = "stun:<myserver>:3478" } 
        },
        .datachannel = DATA_CHANNEL_BINARY, // outgoing type - had to modify sctp_incoming_data to receive this too
        .audio_codec = CODEC_PCMU, 
        .onaudiotrack = webrtc_onAudioTrack,
    };

```

I'm hosting my own server (eturnal) and would rather not disclose its public address. The credentials work, and using STUN works just fine for my environment.

What I believe I observed was that during ICE my ESP32 would send UDP packets directly to the peer (a browser), but not to the TURN server: So, I don't know how the dynamic NAT pathway between the TURN server back to my ESP32 would be established.

To be fair, I haven't fully studied how TURN should work - I just tried to enable it and ICE didn't seem to succeed. If I have your assurance that it should work, then I'll spend some more time on it later.

**Comment by MohamadReza-Rahimi** [comment]
hi thanks for your great library
how can i use this library in espressif ide (eclipse base) with esp32-cam board?
i have downloaded zip file and import in espressif ide but during  building the project i have been receiving many error about depency , makefile and etc 
do need to change these files in eclipse ide (espressif)?
